### DISCUSSION TITLE
Backchannel logout from IdP to my key cloak not working

### DISCUSSION THREAD

**Discussion by TheVaan** [keycloak/keycloak#29158] [Q&A]
Hey there,

we have the following setup:
OurApp -> OurKeycloak -> External-Identity-Provider -> ExternalApp

In cooperation with the owner of the External-Identity-Provider, we want to backchannel logout vice versa through all apps. We activated backchannel-logout for the IDP and in our realms client.
After logging into OurApp by using External-Identity-Provider and then logging into ExternalApp using External-Identity-Provider too, we logout OurApp what results in being logged out of ExternalApp as desired.
But when (after logging into both apps via IDP) and logging out of ExternalApp, we are still logged in in OurApp.

We don't know if *we* misconfigured something or our External-Identity-Provider partner. They asked us for backchannel url and we gave them `https://our-keycloak.url/realms/{our-realm}/protocol/openid-connect/logout/backchannel-logout` as mentioned in [Docs: Keycloak server OIDC URI endpoints](https://www.keycloak.org/docs/latest/server_admin/#con-server-oidc-uri-endpoints_server_administration_guide).

Could someone please help us solving this problem?

**Comment by codespearhead** [comment]
Make sure your client's settings look like this:

![Backchannel logout client config](https://github.com/keycloak/keycloak/assets/72931357/c7a3fb1e-4ff9-4550-8934-a69d1df95972)

Depending on your current environment, you might also want to try as [this comment](https://github.com/keycloak/keycloak/discussions/25684#discussioncomment-7909718) says.

Possibly related issues: #17318 , #25684 , #21770 , #23752 , #9548 .

**Reply by TheVaan** [reply]
Thank you for your fast reply! The config matches your recommendation.

We enabled extended logging as mentioned [here](https://github.com/keycloak/keycloak/discussions/25684#discussioncomment-7909315), but took `DEBUG` instead of `INFO`.

FYI: We are on 24.0.3

This leads us to the following log entries when logging out of ExternalApp:

```
2024-04-30 07:27:33,954 WARN  [org.keycloak.events] (executor-thread-1) type="LOGOUT_ERROR", realmId="xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", clientId="null", userId="null", ipAddress="_xxx.xxx.xxx.xxx_", error="invalid_token"
2024-04-30 07:27:33,956 DEBUG [WebApplicationException] (executor-thread-1) Restarting handler chain for exception exception: org.keycloak.services.ErrorResponseException: HTTP 500 Internal Server Error
	at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint.backchannelLogout(LogoutEndpoint.java:579)
	at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint$quarkusrestinvoker$backchannelLogout_81daba2f3ee97080feca4b7ff4f4bc012c7673c3.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

I'm confused why `clientId` and `userId` are `null`.

**Reply by luigui489** [reply]
@TheVaan Good afternoon.

Did you find the URL that needs to be configured in the IDP in the Backchannel Logout URL field?  I'm having the same problem right now, and authentication via Backchannel Logout isn't working.

Thank you.

**Reply by ittaq** [reply]
Hey @luigui489,

I am answering on behalf of @TheVaan. I am not quite sure what the error was, but I think it was our Identity Provider configuration. Make sure you have set the following fields like in the image. 
![oidc_settings](https://github.com/user-attachments/assets/66759e02-a280-4803-9359-d4256df36be2)
Check the .well-known from the other identity provider and make sure to adjust these endpoints. If your other identity provider also uses Keyloak, then it should have the same structure as in the image. However, check their .well_known.

Referring to the field from above, _Backchannel logout URL._ I cannot say if we entered a value there. We deactivated the login via the other identity provider and are not using it for now...maybe we removed that line. But like now it is empty and the _Backchannel logout session required_ checkbox is selected.

However, I think we only had a misconfiguration regarding the endpoints. Check that in your Keycloak instance. 

If it does not fix your problem, I also recommend starting 2 minimal Keyloak instances locally and try to integrate the second into the first one. That helps you to understand how the systems need to collaborate. This video is pretty old, but it helped me a bit doing that (https://www.youtube.com/watch?v=YIgkIBaTGFE)

**Comment by luigui489** [comment]
Hey @ittaq 

Thanks for your response,  I was able to establish a connection between the two KeyCloaks by configuring the URLs as shown in the image. Authentication through the identity provider works correctly. The problem is that when I log out of the identity provider, the session in the identity broker remains. I think this is more related to the Backchannel URL field. I've tried the URLs https://<brokerhost>/realms/<realm-name>/protocol/openid-connect/logout and https://<brokerhost>/realms/<realm>/protocol/openid-connect/logout/backchannel, but they don't work.

In the case of the URL https://<brokerhost>/realms/<realm-name>/protocol/openid-connect/logout, the logout backchannel causes a POST request to be sent to the broker to log out. However, the broker responds with an HTML request to confirm the logout. In this case, it should not ask for logout confirmation.

https://<brokerhost>/realms/<realm>/protocol/openid-connect/logout/backchannel When using this URL, a BadRequest error occurs. It appears that this API is not suitable for requesting a logout from the identity provider to the broker.

This is the configuration on the client side of the identity provider.

<img width="1028" height="583" alt="image" src="https://github.com/user-attachments/assets/1cef3eac-aed8-4947-97bb-34c9f28835f1" />

**Reply by ittaq** [reply]
Hey @luigui489 ,

did you try these settings as well? In your post you only used /logout and /backchannel. Did you also try backchannel-logout?

Keycloak A -> App A -> Backchannel logout URL: `http://{keycloak**B**:PORT}/realms/{realm-name}/protocol/openid-connect/logout/backchannel-logout`

Keycloak B -> App B -> Backchannel logout URL: `http://{keycloak**A**:PORT}/realms/{realm-name}/protocol/openid-connect/logout/backchannel-logout`

[`URL from docs https://www.keycloak.org/docs/latest/server_admin/#con-server-oidc-uri-endpoints_server_administration_guide`](https://www.keycloak.org/docs/latest/server_admin/#con-server-oidc-uri-endpoints_server_administration_guide)
